src/main/java/org/janelia/saalfeldlab/n5/codec/N5Codecs.java

```diff
@@ -0,0 +1,268 @@
+	public static final DataBlockCodecFactory<double[]> DOUBLE = c -> new DefaultDataBlockCodec<>(DataCodec.DOUBLE_BIG_ENDIAN, DoubleArrayDataBlock::new, c);
+	public static final DataBlockCodecFactory<String[]> STRING = StringDataBlockCodec::new;
+	public static final DataBlockCodecFactory<byte[]>   OBJECT = ObjectDataBlockCodec::new;
+
+	private N5Codecs() {}
+
+	public interface DataBlockCodecFactory<T> {
+
+		DataBlockCodec<T> createDataBlockCodec(Compression compression);
+	}
+
+	/**
+	 * DataBlockCodec for all N5 data types, except STRING and OBJECT
+	 */
+	static class DefaultDataBlockCodec<T> implements DataBlockCodec<T> {
+
+		private final DataCodec<T> dataCodec;
+
+		private final DataBlockFactory<T> dataBlockFactory;
+
+		private final Compression compression;
+
+		DefaultDataBlockCodec(
+				final DataCodec<T> dataCodec,
+				final DataBlockFactory<T> dataBlockFactory,
+				final Compression compression) {
+			this.dataCodec = dataCodec;
+			this.dataBlockFactory = dataBlockFactory;
+			this.compression = compression;
+		}
+
+		@Override
+		public ReadData encode(final DataBlock<T> dataBlock) throws IOException {
+			return ReadData.from(out -> {
+				new ChunkHeader(dataBlock.getSize(), dataBlock.getNumElements()).writeTo(out);
+				compression.encode(dataCodec.serialize(dataBlock.getData())).writeTo(out);
+				out.flush();
+			});
+		}
+
+		@Override
+		public DataBlock<T> decode(final ReadData readData, final long[] gridPosition) throws IOException {
+			try(final InputStream in = readData.inputStream()) {
+				final ChunkHeader header = ChunkHeader.readFrom(in, MODE_DEFAULT, MODE_VARLENGTH);
+				final T data = dataCodec.createData(header.numElements());
+				final int numBytes = header.numElements() * dataCodec.bytesPerElement();
+				final ReadData decompressed = compression.decode(ReadData.from(in), numBytes);
+				dataCodec.deserialize(decompressed, data);
+				return dataBlockFactory.createDataBlock(header.blockSize(), gridPosition, data);
+			}
+		}
+	}
+
+	/**
+	 * DataBlockCodec for N5 data type STRING
+	 */
+	static class StringDataBlockCodec implements DataBlockCodec<String[]> {
+
+		private static final Charset ENCODING = StandardCharsets.UTF_8;
+		private static final String NULLCHAR = "\0";
+
+		private final Compression compression;
+
+		StringDataBlockCodec(final Compression compression) {
+			this.compression = compression;
+		}
+
+		@Override
+		public ReadData encode(final DataBlock<String[]> dataBlock) throws IOException {
+			return ReadData.from(out -> {
+				final String flattenedArray = String.join(NULLCHAR, dataBlock.getData()) + NULLCHAR;
+				final byte[] serializedData = flattenedArray.getBytes(ENCODING);
+				new ChunkHeader(dataBlock.getSize(), serializedData.length).writeTo(out);
+				compression.encode(ReadData.from(serializedData)).writeTo(out);
+				out.flush();
```

Comment:
`out.flush()` is called in all the encode methods here - should it be the responsibility of the `OutputStreamOperator` to flush? could it instead be the job of `LazyReadData`? 

<strike>I'll make a note there too.</strike>

Edit: okay maybe I get it now.  `OutputStreamWriter` is a functional interface and can't be flushed itself.
This leads to a different question.

Reply:
I revised where flush() happens. 👍